base/sparse/sparsematrix.jl

```diff
@@ -613,115 +613,326 @@ function sparse(B::Bidiagonal)
-        throw(DimensionMismatch("the length of the permtuation vector, length(q) = "
-            * "$(length(q)), must match the number of columns of the second argument,"
-            * "A.n = $(An)") )
-    elseif !(length(Crowval) >= Annz)
-        throw(ArgumentError("the first argument's row-index array's length,"
-            * "length(C.rowval) = $(length(Crowval)), must be at least the number of"
-            * "allocated entries in the second argument, nnz(A) = $(Annz)") )
-    elseif !(length(Cnzval) >= Annz)
-        throw(ArgumentError("the first argument's entry-value array's length,"
-            * "length(C.nzval) = $(length(Cnzval)), must be at least the number of"
-            * "allocated entries in the second argument, nnz(A) = $(Annz)") )
-    end
-
-    # Compute the column counts of C and store them shifted forward by one in Ccolptr
-    Ccolptr[1:end] = 0
-    @inbounds for k in 1:Annz
-        Ccolptr[Arowval[k]+1] += 1
-    end
-
-    # From these column counts, compute C's column pointers
-    # and store them shifted forward by one in Ccolptr
+function halfperm!{Tv,Ti<:Integer}(X::SparseMatrixCSC{Tv,Ti}, A::SparseMatrixCSC{Tv,Ti},
+        q::AbstractVector, f = identity)
+    _computecolptrs_halfperm!(X, A)
+    _distributevals_halfperm!(X, A, q, f)
+    return X
+end
+"""
+Helper method for `halfperm!`. Computes `transpose(A[:,q])`'s column pointers, storing them
+shifted one position forward in `X.colptr`; `_distributevals_halfperm!` fixes this shift.
+"""
+function _computecolptrs_halfperm!{Tv,Ti<:Integer}(X::SparseMatrixCSC{Tv,Ti},
+        A::SparseMatrixCSC{Tv,Ti})
+    # Compute `transpose(A[:,q])`'s col counts. Store shifted forward one pos. in X.colptr.
+    fill!(X.colptr, 0)
+    @inbounds for k in 1:nnz(A)
+        X.colptr[A.rowval[k] + 1] += 1
+    end
+    # Compute `transpose(A[:,q])`'s colptrs. Store shifted forward one pos. in X.colptr.
+    X.colptr[1] = 1
     countsum = 1
-    @inbounds for k in 2:(Cn+1)
-        overwritten = Ccolptr[k]
-        Ccolptr[k] = countsum
+    @inbounds for k in 2:(A.m + 1)
+        overwritten = X.colptr[k]
+        X.colptr[k] = countsum
         countsum += overwritten
     end
+end
+"""
+Helper method for `halfperm!`. With `transpose(A[:,q])`'s column pointers shifted one
+position forward in `X.colptr`, computes `map(f, transpose(A[:,q]))` by appropriately
+distributing `A.rowval` and `f`-transformed `A.nzval` into `X.rowval` and `X.nzval`
+respectively. Simultaneously fixes the one-position-forward shift in `X.colptr`.
+"""
+function _distributevals_halfperm!{Tv,Ti<:Integer}(X::SparseMatrixCSC{Tv,Ti},
+        A::SparseMatrixCSC{Tv,Ti}, q::AbstractVector, f)
+    @inbounds for Xi in 1:A.n
+        Aj = q[Xi]
+        for Ak in A.colptr[Aj]:(A.colptr[Aj + 1] - 1)
+            Ai = A.rowval[Ak]
+            Xk = X.colptr[Ai + 1]
+            X.rowval[Xk] = Xi
+            X.nzval[Xk] = f(A.nzval[Ak])
+            X.colptr[Ai + 1] += 1
+        end
+    end
```

Comment:
may want to make sure this returns something, otherwise I think it might be type-unstable if the loop happens to be empty